File: app/src/main/kotlin/com/apex/trader/data/repository/SymbolRules.kt
package com.apex.trader.data.repository

import com.apex.trader.data.api.dto.SymbolDto
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * Per-symbol rounding rules derived from the exchangeInfo `filters`. Binance rejects
 * orders that don't conform to tickSize/stepSize, so we always normalize values
 * client-side before placing an order.
 */
data class SymbolRules(
    val symbol: String,
    val tickSize: Double,
    val stepSize: Double,
    val minQty: Double,
    val minNotional: Double,
    val pricePrecision: Int,
    val quantityPrecision: Int
) {
    fun roundPrice(price: Double): Double = roundToTick(price, tickSize)
    fun roundQuantity(qty: Double): Double = roundToTick(qty, stepSize)
    fun formatPrice(price: Double): String = "%.${pricePrecision}f".format(roundPrice(price))
    fun formatQuantity(qty: Double): String = "%.${quantityPrecision}f".format(roundQuantity(qty))

    private fun roundToTick(value: Double, tick: Double): Double {
        if (tick <= 0.0) return value
        return floor(value / tick) * tick
    }

    companion object {
        fun from(dto: SymbolDto): SymbolRules {
            val priceFilter = dto.filters.firstOrNull { it.filterType == "PRICE_FILTER" }
            val lotFilter = dto.filters.firstOrNull { it.filterType == "LOT_SIZE" }
            val notionalFilter = dto.filters.firstOrNull {
                it.filterType == "MIN_NOTIONAL" || it.filterType == "NOTIONAL"
            }
            val tickSize = priceFilter?.tickSize?.toDoubleOrNull() ?: 10.0.pow(-dto.pricePrecision)
            val stepSize = lotFilter?.stepSize?.toDoubleOrNull() ?: 10.0.pow(-dto.quantityPrecision)
            val minQty = lotFilter?.minQty?.toDoubleOrNull() ?: 0.0
            val minNotional = (notionalFilter?.notional ?: notionalFilter?.minNotional)
                ?.toDoubleOrNull() ?: 5.0
            return SymbolRules(
                symbol = dto.symbol,
                tickSize = tickSize,
                stepSize = stepSize,
                minQty = minQty,
                minNotional = minNotional,
                pricePrecision = dto.pricePrecision.takeIf { it >= 0 } ?: tickPrecision(tickSize),
                quantityPrecision = dto.quantityPrecision.takeIf { it >= 0 } ?: tickPrecision(stepSize)
            )
        }

        private fun tickPrecision(tick: Double): Int {
            if (tick <= 0) return 8
            val p = -ln(tick) / ln(10.0)
            return max(0, p.roundToInt())
        }
    }
}

File: lib/data/models/symbol_rules.dart
import 'dart:math' as math;

/// Per-symbol rounding rules derived from Binance `exchangeInfo`. Orders that
/// don't conform to tickSize/stepSize are rejected, so we always normalize
/// values client-side before placing an order.
class SymbolRules {
  const SymbolRules({
    required this.symbol,
    required this.tickSize,
    required this.stepSize,
    required this.minQty,
    required this.minNotional,
    required this.pricePrecision,
    required this.quantityPrecision,
  });

  final String symbol;
  final double tickSize;
  final double stepSize;
  final double minQty;
  final double minNotional;
  final int pricePrecision;
  final int quantityPrecision;

  double roundPrice(double v) => _floorTo(v, tickSize);
  double roundQuantity(double v) => _floorTo(v, stepSize);
  String formatPrice(double v) => roundPrice(v).toStringAsFixed(pricePrecision);
  String formatQuantity(double v) => roundQuantity(v).toStringAsFixed(quantityPrecision);

  static double _floorTo(double v, double tick) {
    if (tick <= 0) return v;
    return (v / tick).floor() * tick;
  }

  factory SymbolRules.fromJson(Map<String, dynamic> j) {
    final filters = (j['filters'] as List? ?? const []).cast<Map<String, dynamic>>();
    Map<String, dynamic>? f(String type) =>
        filters.firstWhere((m) => m['filterType'] == type, orElse: () => <String, dynamic>{});

    double parseD(Object? v, double fallback) {
      if (v == null) return fallback;
      return double.tryParse(v.toString()) ?? fallback;
    }

    int parseI(Object? v, int fallback) {
      if (v == null) return fallback;
      return int.tryParse(v.toString()) ?? fallback;
    }

    final price = f('PRICE_FILTER');
    final lot = f('LOT_SIZE');
    final notional = filters.firstWhere(
      (m) => m['filterType'] == 'MIN_NOTIONAL' || m['filterType'] == 'NOTIONAL',
      orElse: () => <String, dynamic>{},
    );

    final pp = parseI(j['pricePrecision'], 2);
    final qp = parseI(j['quantityPrecision'], 3);
    final tick = parseD(price?['tickSize'], math.pow(10, -pp).toDouble());
    final step = parseD(lot?['stepSize'], math.pow(10, -qp).toDouble());
    final minQ = parseD(lot?['minQty'], 0);
    final minN = parseD(notional['notional'] ?? notional['minNotional'], 5);

    return SymbolRules(
      symbol: j['symbol'] as String,
      tickSize: tick,
      stepSize: step,
      minQty: minQ,
      minNotional: minN,
      pricePrecision: pp,
      quantityPrecision: qp,
    );
  }
}
